tonic/src/transport/service/user_agent.rs

```diff
@@ -0,0 +1,47 @@
+use http::{header::USER_AGENT, HeaderValue, Request};
+use std::task::{Context, Poll};
+use tower_service::Service;
+
+const TONIC_USER_AGENT: &str = concat!("tonic/", env!("CARGO_PKG_VERSION"));
+
+#[derive(Debug)]
+pub(crate) struct UserAgent<T> {
+    inner: T,
+    user_agent: Option<HeaderValue>,
+}
+
+impl<T> UserAgent<T> {
+    pub(crate) fn new(inner: T, user_agent: Option<HeaderValue>) -> Self {
+        Self { inner, user_agent }
+    }
+}
+
+impl<T, ReqBody> Service<Request<ReqBody>> for UserAgent<T>
+where
+    T: Service<Request<ReqBody>>,
+{
+    type Response = T::Response;
+    type Error = T::Error;
+    type Future = T::Future;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, mut req: Request<ReqBody>) -> Self::Future {
+        let value = match self.user_agent.as_ref() {
+            Some(custom_ua) => {
```

Comment:
can we compute this at `fn new` time then just clone the headervalue everytime, instead of reallocating on each request?

Reply:
Good idea.